auth/auth/auth.py

```diff
@@ -0,0 +1,233 @@
+import logging
+import aiohttp
+from aiohttp import web
+import aiohttp_session
+import uvloop
+
+import google.auth.transport.requests
+import google.oauth2.id_token
+import google_auth_oauthlib.flow
+
+from hailtop.gear import get_deploy_config, setup_aiohttp_session, create_database_pool
+from hailtop.gear.auth import rest_authenticated_users_only, \
+    web_maybe_authenticated_user, create_session
+
+log = logging.getLogger('auth')
+
+uvloop.install()
+
+deploy_config = get_deploy_config()
+
+routes = web.RouteTableDef()
+
+
+def get_flow(redirect_uri, state=None):
+    scopes = [
+        'https://www.googleapis.com/auth/userinfo.profile',
+        'https://www.googleapis.com/auth/userinfo.email',
+        'openid'
+    ]
+    flow = google_auth_oauthlib.flow.Flow.from_client_secrets_file(
+        '/auth-oauth2-client-secret/client_secret.json', scopes=scopes, state=state)
+    flow.redirect_uri = redirect_uri
+    return flow
+
+
+@routes.get('/healthcheck')
+async def get_healthcheck(request):  # pylint: disable=W0613
+    return web.Response()
+
+
+@routes.get('')
+@routes.get('/')
+async def get_index(request):  # pylint: disable=unused-argument
+    return aiohttp.web.HTTPFound('/login')
+
+
+@routes.get('/login')
+@web_maybe_authenticated_user
+async def login(request, userdata):
+    next = request.query.get('next', deploy_config.external_url('notebook2', ''))
+    if userdata:
+        return aiohttp.web.HTTPFound(next)
+
+    flow = get_flow(deploy_config.external_url('auth', '/oauth2callback'))
+
+    authorization_url, state = flow.authorization_url(
+        access_type='offline',
+        include_granted_scopes='true')
+
+    session = await aiohttp_session.get_session(request)
+    session['state'] = state
+    session['next'] = next
+
+    return aiohttp.web.HTTPFound(authorization_url)
+
+
+@routes.get('/oauth2callback')
+async def callback(request):
+    session = await aiohttp_session.get_session(request)
+    if 'state' not in session:
+        raise web.HTTPUnauthorized()
+    state = session['state']
+
+    flow = get_flow(deploy_config.external_url('auth', '/oauth2callback'), state=state)
+
+    try:
+        flow.fetch_token(code=request.query['code'])
+        token = google.oauth2.id_token.verify_oauth2_token(
+            flow.credentials.id_token, google.auth.transport.requests.Request())
+        id = token['sub']
+    except Exception:
+        log.exception('oauth2 callback: could not fetch and verify token')
+        raise web.HTTPUnauthorized()
+
+    dbpool = request.app['dbpool']
+    async with dbpool.acquire() as conn:
+        async with conn.cursor() as cursor:
+            await cursor.execute('SELECT * from user_data where user_id = %s;', f'google-oauth2|{id}')
```

Comment:
This is right, but I want to make sure that we don't want to provision for multiple login methods tied to a single account. If we did want that it may be better not to have a single user_id, but google_id, etc. Downside there is schema change for a new connection, but that doesn't seem like a big performance issue, unless we have a billion users. Or, we can go schema less  route, and use JSON_EXTRACT, but that seems more complicated. If first route (multiple user_id's), this is down to a naming issue (google_user_id).

Reply:
I'm happy to just support Google now, and have one Google account per Hail account.

I would say we shouldn't try to envision every potential use case now, but rather make this code easy to change in the future (e.g. by making database migrations easy, something on the todo list).

Reply:
Additionally, I think we should implement a refresh mechanism, maybe as a step 2. Every time the user visits the site with a valid session, we should mark their last visit date, and adjust the session expiration date based on that. This could also allow us to use much shorter sessions. To help prevent a malicious use of this, could set a flag on the session, that states whether it can be rehydrated. In particular, if I visit the site with a valid session (meaning I see my username/profile), I expect that the next action I take is also valid. Our current mechanism will have users that see a perfectly valid page one second, and then get a login redirect the next.